js/web/lib/wasm/jsep/webgpu/ops/conv-transpose.ts

```diff
@@ -0,0 +1,257 @@
+  if (padsSet && attributes.pads.length !== spatialRank * 2) {
+    throw new Error(`pads should be ${spatialRank * 2}D`);
+  }
+
+  // Wrong output padding dimension
+  if (attributes.outputPadding.length !== spatialRank && attributes.outputPadding.length !== 0) {
+    throw new Error(`output_padding should be ${spatialRank}D`);
+  }
+
+  // if kernelShape is specified, it's data length must be 2 less than dims length of the weights tensor
+  // (the first 2 dims are batch_size and channels)
+  const kernelShapeSet = attributes.kernelShape.reduce((a, b) => a + b, 0) > 0;
+  if (kernelShapeSet && attributes.kernelShape.length !== 0 &&
+      attributes.kernelShape.length !== inputs[1].dims.length - 2) {
+    throw new Error('invalid kernel shape');
+  }
+
+  // as with kernelShape, must have same number of spatial dims as input
+  if (attributes.outputShape.length !== 0 && attributes.outputShape.length !== inputs[0].dims.length - 2) {
+    throw new Error('invalid output shape');
+  }
+
+  // TODO : Need to add support for float64
+  if (inputs[0].dataType !== DataType.float || inputs[1].dataType !== DataType.float) {
+    throw new Error('ConvTranspose input(X,W) should be float tensor');
+  }
+
+  if (inputs.length === 3 && inputs[2].dataType !== DataType.float) {
+    throw new Error('ConvTranspose input(bias) should be float tensor');
+  }
+};
+
+
+const convTranspose2d =
+    (context: ComputeContext, inputs: readonly TensorView[], attributes: ConvTransposeAttributes): void => {
+      const adjustedAttributes = getAdjustedConvTransposeAttributes(attributes, inputs);
+
+      const hasBias = inputs.length === 3;
+      // const hasPreluActivationWeights = false; /* TODO: add support for prelu activation weights */
+      const isChannelsLast = adjustedAttributes.format === 'NHWC';
+
+      const outputShape = adjustedAttributes.outputShape;
+
+      // STEP.1: transpose weight
+      const transposedWeight = (context.customData.wT as TensorView | undefined) ??
+          context.compute(
+              {
+                ...transposeProgramMetadata,
+                cacheHint: weightTransposeAttribute.cacheKey,
+                get: () => createTransposeProgramInfo(inputs[1], weightTransposeAttribute.perm)
+              },
+              {inputs: [1], outputs: [attributes.wIsConst ? -2 : -1]})[0];
+      if (attributes.wIsConst && !context.customData.wT) {
+        context.customData.wT = transposedWeight;
+      }
+
+      // STEP.2: prepare reshaped inputs
+      const convInputs = [inputs[0], transposedWeight];
+      if (hasBias) {
+        if (!isChannelsLast && inputs[2].dims.length === 1) {
+          convInputs.push(inputs[2].reshape([inputs[2].dims[0], 1, 1]));
+        } else {
+          convInputs.push(inputs[2]);
+        }
+      }
+
+      // STEP.3: compute matmul
+      context.compute(
+          createConvTranspose2DProgramInfoLoader(convInputs, adjustedAttributes, outputShape, hasBias),
+          {inputs: convInputs});
```

Comment:
the current implementation depends on a pre-transposed weight, which is not "naive". A native is expected to be as simple as possible ( mostly is a brute-force implementation ).

and the file organization is confusing. the current files look like there are 2 different implementations - one is an optimized on defined in conv-transpose.ts and the other is a simple one defined in conv-transpose-naive.ts. it turns out that there is only one implementation. I suggest to merge those 2 files into one unless we really get 2 different implementations.

Reply:
I have created conv-transpose-naive.ts only as a wrapper to the implementation in 3rd-pary/conv-backprop-webgpu.ts along the same lines as conv2d-mm.ts. Does changing the name from conv-transpose-naive.ts to conv-transpose2d-naive.ts make it clear? My idea is that we eventually add a MatMul version also.